[쇼핑 장바구니 1, 2단계] 이든 미션 제출합니다.

devfeijoa opened the pull request:
안녕하세요, 둘리 ! 이든 입니다.
귀중한 시간내어 리뷰해주셔서 감사합니다 ☺️


# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

테스트 코드 작성과 리팩토링을 하지 못했습니다 😢 
빠진 부분 빠르게 작성하여 다음 리뷰 요청때 반영하여 올리도록 하겠습니다 !

이전 미션에서부터 배워야 할 학습이 덜 된채로 미션이 진행되고 있어서 부족한 부분이 많습니다.
안드로이드 경험이 없어서 이해가 많이 어려운 상태라 피드백을 단계별로 부탁드려도 될까요? 😢 
늘 감사합니다 !



 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
-

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

https://github.com/user-attachments/assets/66adf6b2-5387-493a-ad17-598db9ace91f


 <!-- 필요하다면 스크린샷을 첨부하세요. -->

[](url)
![]()

parkjiminnnn committed:
docs: 기능 명세 추가

parkjiminnnn committed:
build: 데이터바인딩 의존성 추가

parkjiminnnn committed:
feat: 상품 목록 뷰 구현

devfeijoa committed:
feat: toolbar 구현

parkjiminnnn committed:
refactor: bindingAdapter, Glide 적용

parkjiminnnn committed:
feat: 장바구니 뷰 구현

parkjiminnnn committed:
refactor: 상품목록 뷰 개선

parkjiminnnn committed:
feat: 상품 상세 뷰 구현

parkjiminnnn committed:
chore: 이미지 파일 제거

parkjiminnnn committed:
chore: 패키지 구조 변경

parkjiminnnn committed:
feat: 상품 상세화면 이동기능 구현

parkjiminnnn committed:
feat: 장바구니에 담긴 상품 뷰 구현

parkjiminnnn committed:
feat: 장바구니 담기 기능 구현

parkjiminnnn committed:
feat: 장바구니 상품 제거 기능 구현

parkjiminnnn committed:
feat: CartViewModel 생성

parkjiminnnn committed:
refactor: 장바구니 상품 추가 기능 변경

parkjiminnnn committed:
feat: 장바구니 상품 제거 기능 임시 구현

devfeijoa committed:
fix: 장바구니 아이템 추가,삭제 기능 개선

parkjiminnnn committed:
refactor: 화면 이동 경로 변경

parkjiminnnn committed:
refactor: 함수분리

parkjiminnnn committed:
refactor: 함수명 변경

parkjiminnnn committed:
feat: ProductsViewModel 생성

parkjiminnnn committed:
feat: 페이지 이동 버튼 뷰 구현

parkjiminnnn committed:
feat: ProductDetailViewModel 구현

parkjiminnnn committed:
feat: 페이지 이동 기능 구현

parkjiminnnn committed:
feat: 더보기 버튼 기능 구현

parkjiminnnn committed:
fix: 더보기 버튼이 사라지지 않는 오류 수정

parkjiminnnn committed:
docs: 기능 명세 추가

hyemdooly commented on app/src/main/AndroidManifest.xml:
```diff
@@ -13,14 +15,23 @@
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
         </activity>
+        <activity
+            android:name=".MainActivity"
+            android:exported="true" />
```
MainActivity는 사용하지 않으니 삭제할까요?

devfeijoa replied:
해당 리뷰 반영하여 수정하였습니다 ! 
6548708f86d136567600b82c764346d3a39a6363

hyemdooly commented on app/src/main/java/woowacourse/shopping/model/intent/IntentExtension.kt:
```diff
@@ -0,0 +1,15 @@
+import android.os.Build
+import java.io.Serializable
+
+inline fun <reified T : Serializable> Intent.getSerializableExtraData(key: String): T? =
```
보통 이렇게 버전별로 호환시키기 위한 함수 이름은 `XXXCompat`이라고 짓고는 합니다!
inline과 Generic, refied 키워드에 대해서도 시간 여유가 된다면 학습해보시죠!

devfeijoa replied:
해당 리뷰 반영하여 수정하였습니다 ! 
cacb6052fb4c07923f8fadfe482ebbde54a41451

hyemdooly commented on app/src/main/java/woowacourse/shopping/model/products/Products.kt:
```diff
@@ -0,0 +1,161 @@
+package woowacourse.shopping.model.products
+
+data class Products(
+    val value: List<Product> = dummyProducts,
```
더미데이터는 모델에서 아예 분리했으면 좋겠어요!

devfeijoa replied:
해당 리뷰 반영하여 수정하였습니다 !
9ce0b30b568bed4d7cfd28ba2eb39939cb81066f

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/adapter/BindingAdapter.kt:
```diff
@@ -0,0 +1,16 @@
+import androidx.databinding.BindingAdapter
+import com.bumptech.glide.Glide
+
+@BindingAdapter("app:imageUrl")
+fun setImageUrl(
+    imageView: ImageView,
+    imageUrl: String,
+) {
+    Glide
+        .with(imageView.context)
+        .load(imageUrl)
+        .into(imageView)
+}
```
바인딩어댑터 활용 좋습니다 👍 
`app:`은 사용해준 이유가 있으신가요?

devfeijoa replied:
`app: ` 네임 스페이스를 붙여서 커스텀한 값임을 명확히 하려했습니다 ! android가 아닌 app을 붙여서 안드로이드 프레임워크에서 기본으로 제공하는 속성이 아님을 분명히 하고자 했습니다 !

hyemdooly replied:
`app:`을 안붙여도 `app:imageUrl`로 사용할 수 있던 것 같아서 코멘트 드렸어요 ㅎㅎ

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/CartActivity.kt:
```diff
@@ -0,0 +1,89 @@
+    private lateinit var adapter: CartAdapter
+    private val viewModel: CartViewModel by viewModels { CartViewModel.Factory }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
```
`onCreate` 함수가 기네요! observe 코드들을 분리할 수 있을까요?

hyemdooly replied:
요것도 같이 반영해주세요!

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/CartActivity.kt:
```diff
@@ -0,0 +1,89 @@
+            binding.tvPageCount.text = pageCount.toString()
+            if (viewModel.isFirstPage(pageCount)) {
+                binding.btnPreviousPage.setBackgroundColor(
+                    ContextCompat.getColor(
```
ContextCompat의 getColor를 사용하지 않아도, getColor를 바로 사용할 수 있어요!

devfeijoa replied:
`ContextCompat.getColor()` 대신 `Context.getColor()`를 사용하도록 변경하였습니다 !

둘리의 피드백을 보고 찾아보니, API 레벨 23 이상에서 `Context.getColor()`가 직접 지원되므로 더 이상 `ContextCompat`를 사용할 필요가 없다고 합니다 !

따라서 `binding.btnPreviousPage.context.getColor()`와 같이 뷰의 컨텍스트를 사용하여 색상을 가져오도록 수정했습니다 !

bc7805e971f76fbd756af64bc035facef9cd5a5f

hyemdooly replied:
바로 `getColor`함수를 사용할 수도 있습니다 ㅎㅎ 왜인지도 살펴보시죠!

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/CartAdapter.kt:
```diff
@@ -0,0 +1,42 @@
+import woowacourse.shopping.model.products.Product
+
+class CartAdapter(
+    private val itemsInCart: MutableList<Product> = mutableListOf(),
```
이렇게 MutableList로 받으면 어떤 문제가 발생할 수 있을까요?

devfeijoa replied:
외부에서 데이터 변경이 가능하다는 문제가 생길수 있습니다 ! 또한 이렇게 외부에서 리스트를 변경했을때 어댑터가 그 변경사항을 즉시 알 수 없다는 문제가 생길 수 있습니다..!

따라서 다음 리뷰요청에는 불변객체로 수정하도록 하겠습니다 !

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/CartViewHolder.kt:
```diff
@@ -0,0 +1,15 @@
+    private val onProductRemoveClickListener: OnProductRemoveClickListener,
+) : RecyclerView.ViewHolder(binding.root) {
+    fun bind(item: Product) {
+        binding.onProductRemoveClickListener = onProductRemoveClickListener
```
처음 생성할 때 받고 변하지 않으니 초기화할 때만 대입해줘도 될 것 같아요~

hyemdooly commented on app/src/main/java/woowacourse/shopping/viewmodel/cart/CartViewModel.kt:
```diff
@@ -0,0 +1,90 @@
+                    extras: CreationExtras,
+                ): T {
+                    checkNotNull(extras[APPLICATION_KEY])
+                    extras.createSavedStateHandle()
```
`createSavedStateHandle`은 어떤 함수일까요?

devfeijoa replied:
`createSaveStateHandle()`은 ViewModel이 프로세스 종료 후에도 데이터를 보존하고 복원할 수 있도록 도와줍니다.

SavedStateHandle은 Bundle처럼 작동하며, 액티비티나 프래그먼트가 재생성될 때(예: 프로세스 종료 후 복원되거나, 화면 회전 등으로 인해) 상태를 자동으로 저장하고 복원해 줍니다..!

하지만 제 코드에서는 CartViewModel의 생성자로 이 SavedStateHandle이 전달되고 있지 않습니다..! 따라서 SaveStateHandle을 생성하고, CartViewModel 에서 상태복원에 활용하라고 피드백을 주셨던걸까요...?
아직 잘 이해가 안되서 최대한 이해한 것을 바탕으로 질문드립니다 !

hyemdooly replied:
내부 코드를 이해하고 사용했는지, 팩토리 함수 전체적으로 필요한 로직이었는지, 이든의 의도대로 활용되고 있는지 의문이 들어 드린 코멘트였습니다!

hyemdooly commented on app/src/main/java/woowacourse/shopping/viewmodel/products/ProductsViewModel.kt:
```diff
@@ -0,0 +1,59 @@
+import woowacourse.shopping.model.products.Products
+
+class ProductsViewModel(
+    private val products: Products,
```
상품을 전부 받고 잘라서 사용하는 것은 이번 미션 페이지네이션에 이점이 없다고 생각해요!
만약에 몇백, 몇천개의 products가 존재한다면 어떨까요? 😱 그 안에서 nn개씩 리스트를 잘라서 붙이는 로직을 계속해서 실행해야할거에요.

이제 repository 패턴을 학습하고 적용해보아요!
Room은 사용하지 않으셔도 괜찮고, 지금 Cart object처럼 MutableList를 관리하는 Repository 클래스를 하나 만든다고 생각해보아요. (Cart쪽도 동일하게 바꿔주세요!)
그 안에서 상품을 원하는만큼만 리턴해주는 함수를 만들어보면 어떨까요?

물론 이렇게 구현한다면, 지금 구현에서 로직만 분리한 것과 다른게 있을까? 🤔 라고 생각할 수 있습니다.
하지만 이렇게 분리해둔다면 나중에 서버를 붙일 때나, Room을 붙일 때 요 Repository만 수정하거나 새로운 구현체를 만들어주면 쉽게 할 수 있어요!

devfeijoa replied:
해당 리뷰 반영하여 Repository 패턴을 적용해보았습니다..!
Cart쪽은 아직 반영하지 못했는데, 다음 리뷰 요청때는 Cart쪽도 Repository 패턴을 반영하여 요청드리겠습니다 !

4afdf2573701a67c83ef3e6a6f90ba90068f0b59
ed2aec48812bf49beab5bd5aa0135dc19637167b
아래 커밋은 get으로 시작하는 메서드명을 수정한 커밋입니다

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/products/ProductsActivity.kt:
```diff
@@ -0,0 +1,95 @@
+        }
+
+        binding.cartImageBtn.setOnClickListener {
+            startActivity(Intent(this, CartActivity::class.java))
```
CartActivity가 Intent를 생성하는 책임을 갖게 할 수 있을까요?

devfeijoa replied:
CartActivity가 Intent를 생성하는 책임을 갖도록 팩토리 메서드를 사용하여 개선하였습니다 ! 

2ddad4768426dc35668f0b7a9a5f0730846f175c

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/products/ProductsActivity.kt:
```diff
@@ -0,0 +1,95 @@
+            startActivity(Intent(this, CartActivity::class.java))
+        }
+        binding.rvProducts.addItemDecoration(GridSpacingItemDecoration(SPAN_COUNT, SPACING_DP))
+        binding.rvProducts.addOnScrollListener(
+            object : OnScrollListener() {
+                override fun onScrolled(
+                    recyclerView: RecyclerView,
+                    dx: Int,
+                    dy: Int,
+                ) {
+                    super.onScrolled(recyclerView, dx, dy)
+                    val layoutManager = recyclerView.layoutManager as GridLayoutManager
+                    val lastVisibleItemPosition: Int = layoutManager.findLastVisibleItemPosition()
+                    val totalItemCount: Int = adapter.itemCount
+                    val moreBtn = binding.btnMore
+
+                    if (lastVisibleItemPosition >= totalItemCount - 1) {
+                        val isBottom = viewModel.isAllProductsFetched.value ?: false
+                        if (isBottom) {
+                            binding.btnMore.visibility = GONE
+                        } else {
+                            moreBtn.visibility = VISIBLE
+                        }
+                        moreBtn.setOnClickListener {
+                            it.visibility = GONE
+                            viewModel.loadNextPage()
+                        }
+                    } else {
+                        moreBtn.visibility = GONE
+                    }
+                }
+            },
+        )
```
onScrollListener에서 더보기 visible 처리를 해주고 있네요!
이렇게 되면 스크롤을 할 때마다 이 로직들이 실행될거에요.
상품 목록 + 더보기 버튼은 멀티뷰 리사이클러뷰로 구현해볼 수 있습니다.

devfeijoa replied:
상품목록 + 더보기 버튼을 멀티뷰 리사이클러뷰로 수정하여 구현했습니다 ! 

0efcaa7961bfa162f25a6ba34423094478f62055

hyemdooly commented on app/src/main/res/layout/activity_cart.xml:
```diff
@@ -0,0 +1,102 @@
+            android:layout_height="0dp"
+            android:paddingHorizontal="18dp"
+            android:paddingVertical="24dp"
+            app:layoutManager="androidx.recyclerview.widget.GridLayoutManager"
```
카트는 열이 1인 vertical이기 때문에, LinearLayoutManager를 사용해도 됩니다~!

devfeijoa replied:
해당 피드백 반영하여 수정하였습니다 ! 

9ce08659cad08c7ed8a4fcd3946fd06ebc12a035

hyemdooly reviewed (changes requested):
이든 안녕하세요!! 리뷰가 늦어졌는데 양해해주셔서 감사해요! 🙏 
안드로이드가 처음이시군요. 이번 레벨2가 많이 어려우시겠어요! 😭 
단계별 리뷰를 요청주셔서 코멘트 수를 조절했는데요, 이번에는 repository 패턴을 학습하고 pagination을 개선하는 것과, ScrollListener쪽에 집중해보면 좋겠습니다. 멀티뷰 리사이클러뷰말고 다른 방법이 생각나신다면 다른 방법으로 구현하셔도 괜찮습니다.
참고로 이번 1, 2단계 미션은 아무리 늦어도 목요일 안에는 merge하는 것이 좋아보입니다.
궁금한 점은 DM 또는 코멘트 언제든지 남겨주세요! 💪

devfeijoa committed:
remove: 사용하지 않는 activity 제거

devfeijoa committed:
refactor: 확장 함수 네이밍 변경

devfeijoa committed:
refactor: 더미 데이터 분리

devfeijoa committed:
remove: 사용하지 않는 파일 제거

devfeijoa committed:
refactor: strings.xml 로 변경

devfeijoa committed:
feat: 상품 목록 더보기 구현

- 상품 목록 아이템에 '더보기' 버튼 레이아웃 추가
- 상품 아이템 클릭 리스너 바인딩 추가

devfeijoa committed:
feat: 상품 목록 클릭 리스너 추가

- LoadMoreClickListener: 더보기 버튼 클릭 리스너
- ProductListViewType: 상품 목록 아이템 타입 (상품, 더보기)
- ProductsClickListener: 상품 아이템 클릭 리스너

devfeijoa committed:
feat: LoadMoreViewHolder 추가

상품 목록에서 더 보기 기능을 위한 ViewHolder를 추가합니다.

devfeijoa committed:
feat: 상품 목록 더보기 구현

- 상품 목록 화면에 더보기 버튼을 추가하여 더 많은 상품을 로드할 수 있도록 구현
- `ProductsAdapter`를 멀티 뷰 타입으로 수정하여 상품 아이템과 더보기 버튼 아이템을 구분하여 표시
- `ProductsViewModel`에서 페이지네이션 로직을 수정하여 더보기 버튼의 가시성을 관리

devfeijoa committed:
feat: ProductDetailActivity Intent 생성자 추가

ProductDetailActivity를 시작하는 Intent를 생성하는 newIntent 정적 팩토리 메서드를 추가했습니다.
이를 통해 Product 객체를 Intent에 담아 전달할 수 있습니다.

devfeijoa committed:
feat: 상품 목록 저장소 구현

devfeijoa committed:
Merge branch 'step1' of https://github.com/devfeijoa/android-shopping-cart into step1

devfeijoa committed:
refactor: ProductRepository 메서드명 변경

- `getAllProducts` -> `fetchAllProducts`
- `getSize` -> `productCount`
- `getSinglePage` -> `pageOfProducts`

devfeijoa committed:
fix: 장바구니 화면 레이아웃 매니저 변경

GridLayoutManager에서 LinearLayoutManager로 변경

devfeijoa committed:
refactor: 상품 클릭시 상세페이지로 가도록 개선

devfeijoa committed:
refactor: Context.getColor()를 사용하여 개선

`ContextCompat.getColor()` 대신 `Context.getColor()`를 사용하도록 변경

devfeijoa committed:
refactor: CartActivity와 ProductsActivity에 newIntent 팩토리 메서드로 개선

hyemdooly commented on app/src/main/res/layout/item_load_more.xml:
```diff
@@ -0,0 +1,40 @@
+            name="loadMoreClickListener"
+            type="woowacourse.shopping.view.products.LoadMoreClickListener" />
+
+        <variable
+            name="vm"
+            type="woowacourse.shopping.viewmodel.products.ProductsViewModel" />
+
```
요 리뷰 확인해주세요!

hyemdooly commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,8 @@
     <string name="app_name">Shopping</string>
+    <string name="button_add_to_cart">장바구니 담기</string>
+
+    <string name="product_price_format">%,d원</string>
+    <string name="add_to_cart_message">"장바구니에 담겼습니다"</string>
```
장바구니에 담겼습니다 문자열이 의도된게 맞는지 확인해주세요!

hyemdooly commented on app/src/main/java/woowacourse/shopping/repository/ProductRepositoryImpl.kt:
```diff
@@ -0,0 +1,17 @@
+
+    override fun productCount(): Int = storage.dummyProducts.size
+
+    override fun pageOfProducts(
+        start: Int,
+        end: Int,
+    ): List<Product> = storage.dummyProducts.subList(start, end)
```
start, end 보다는 start, count와 같이 시작과 갯수를 입력받는 것은 어떨까요?
만약 subList에서 범위값 그 이상을 end값으로 주게된다면 어떻게 될까요?

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/productdetail/ProductDetailActivity.kt:
```diff
@@ -44,4 +46,14 @@ class ProductDetailActivity : AppCompatActivity() {
+            product: Product,
+        ): Intent =
+            Intent(context, ProductDetailActivity::class.java).apply {
+                putExtra("product", product)
```
요런 키 값도 상수화를 해주면 좋습니다.

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/products/ProductListViewType.kt:
```diff
@@ -0,0 +1,11 @@
+
+import woowacourse.shopping.model.products.Product
+
+sealed class ProductListViewType {
```
이 sealed class는 item을 의미하는 것 같은데, ~Type이라는 네이밍은 item 객체보다는 enum class와 같은 이름처럼 느껴진달까요 🤔 
더 직관적으로 수정해볼 수 있을까요?

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/products/ProductsActivity.kt:
```diff
@@ -86,9 +58,31 @@ class ProductsActivity : AppCompatActivity() {
             insets
         }
+
+        binding.rvProducts.layoutManager =
+            GridLayoutManager(this, 2).apply {
+                spanSizeLookup =
+                    object : SpanSizeLookup() {
+                        override fun getSpanSize(position: Int): Int {
+                            val viewType = binding.rvProducts.adapter?.getItemViewType(position)
+                            return when (viewType) {
+                                R.layout.item_product -> 1
+                                else -> 2
+                            }
+                        }
+                    }
+            }
```
간단한 숫자도 상수화를 해주시면 좋아요!

hyemdooly commented on app/src/main/java/woowacourse/shopping/viewmodel/products/ProductsViewModel.kt:
```diff
@@ -7,35 +7,42 @@ import androidx.lifecycle.ViewModelProvider
 
-        if (nextStart < products.value.size) {
-            val nextItems = products.value.subList(nextStart, nextEnd)
+        if (nextStart < productRepository.productCount()) {
```
repository로부터 count를 반환해서 사용하는 것보다, repository에게 특정 인덱스로부터 특정 개수를 리턴하게끔 구조를 작성해볼 수 있을 것 같아요!

hyemdooly reviewed (approved):
이든 안녕하세요~! 리뷰 반영하느라 고생 많으셨습니다.
다음 3, 4단계 미션 진행을 위해 이번 1, 2단계 미션은 여기까지 하고 merge하겠습니다.
3, 4단계 미션 진행하면서 남겨드린 리뷰 반영을 함께 해주세요!
다음 미션도 화이팅입니다~! 💪
